File: ios/Sources/App/Models/GitDiff.swift
import Foundation

/// A minimal unified-diff parser — just enough of `git diff`'s output
/// format to drive a colored, per-file review UI. Not a general-purpose
/// diff/patch library; doesn't handle every edge case `git diff` can emit
/// (e.g. combined diffs from a merge), only the common single-parent case.
struct GitDiffFile: Identifiable {
    let id = UUID()
    let oldPath: String
    let newPath: String
    let isNew: Bool
    let isDeleted: Bool
    let isRenamed: Bool
    let isBinary: Bool
    let hunks: [GitDiffHunk]

    var displayPath: String { newPath == "/dev/null" ? oldPath : newPath }
    var additions: Int { hunks.reduce(0) { $0 + $1.lines.filter { $0.kind == .addition }.count } }
    var deletions: Int { hunks.reduce(0) { $0 + $1.lines.filter { $0.kind == .deletion }.count } }
}

struct GitDiffHunk: Identifiable {
    let id = UUID()
    let header: String
    let lines: [GitDiffLine]
}

struct GitDiffLine: Identifiable {
    enum Kind { case context, addition, deletion }
    let id = UUID()
    let kind: Kind
    let text: String
}

enum GitDiff {
    static func parse(_ raw: String) -> [GitDiffFile] {
        guard !raw.isEmpty else { return [] }
        var files: [GitDiffFile] = []

        var oldPath = ""
        var newPath = ""
        var isNew = false
        var isDeleted = false
        var isRenamed = false
        var isBinary = false
        var hunks: [GitDiffHunk] = []
        var currentHunkHeader: String?
        var currentHunkLines: [GitDiffLine] = []
        var hasFile = false

        func flushHunk() {
            if let header = currentHunkHeader {
                hunks.append(GitDiffHunk(header: header, lines: currentHunkLines))
            }
            currentHunkHeader = nil
            currentHunkLines = []
        }

        func flushFile() {
            flushHunk()
            if hasFile {
                files.append(GitDiffFile(
                    oldPath: oldPath, newPath: newPath,
                    isNew: isNew, isDeleted: isDeleted, isRenamed: isRenamed, isBinary: isBinary,
                    hunks: hunks
                ))
            }
            oldPath = ""; newPath = ""
            isNew = false; isDeleted = false; isRenamed = false; isBinary = false
            hunks = []
            hasFile = false
        }

        for line in raw.split(separator: "\n", omittingEmptySubsequences: false) {
            let text = String(line)
            if text.hasPrefix("diff --git ") {
                flushFile()
                hasFile = true
                // "diff --git a/old b/new" — paths themselves can contain
                // spaces, so this split is best-effort; --- / +++ below
                // give the authoritative paths for the common case.
                let parts = text.dropFirst("diff --git ".count).split(separator: " ")
                if parts.count >= 2 {
                    oldPath = String(parts[0].dropFirst(2))
                    newPath = String(parts[1].dropFirst(2))
                }
            } else if text.hasPrefix("new file mode") {
                isNew = true
            } else if text.hasPrefix("deleted file mode") {
                isDeleted = true
            } else if text.hasPrefix("rename from ") {
                isRenamed = true
                oldPath = String(text.dropFirst("rename from ".count))
            } else if text.hasPrefix("rename to ") {
                isRenamed = true
                newPath = String(text.dropFirst("rename to ".count))
            } else if text.hasPrefix("Binary files ") || text.contains("GIT binary patch") {
                isBinary = true
            } else if text.hasPrefix("--- ") {
                let path = String(text.dropFirst(4))
                if path != "/dev/null" { oldPath = path.hasPrefix("a/") ? String(path.dropFirst(2)) : path }
            } else if text.hasPrefix("+++ ") {
                let path = String(text.dropFirst(4))
                if path != "/dev/null" { newPath = path.hasPrefix("b/") ? String(path.dropFirst(2)) : path }
            } else if text.hasPrefix("@@ ") {
                flushHunk()
                currentHunkHeader = text
            } else if currentHunkHeader != nil {
                if text.hasPrefix("+") {
                    currentHunkLines.append(GitDiffLine(kind: .addition, text: String(text.dropFirst())))
                } else if text.hasPrefix("-") {
                    currentHunkLines.append(GitDiffLine(kind: .deletion, text: String(text.dropFirst())))
                } else if text.hasPrefix("\\ ") {
                    // "\ No newline at end of file" — not a real line, skip.
                } else {
                    currentHunkLines.append(GitDiffLine(kind: .context, text: String(text.dropFirst(min(1, text.count)))))
                }
            }
        }
        flushFile()
        return files
    }
}

File: android/app/src/main/kotlin/dev/termvault/app/git/GitDiff.kt
package dev.termvault.app.git

import java.util.UUID

/** Direct Kotlin port of iOS `GitDiff.swift` — a dependency-free unified-diff parser. */

data class GitDiffLine(
    val id: UUID = UUID.randomUUID(),
    val kind: Kind,
    val text: String,
) {
    enum class Kind { CONTEXT, ADDITION, DELETION }
}

data class GitDiffHunk(
    val id: UUID = UUID.randomUUID(),
    val header: String,
    val lines: List<GitDiffLine>,
)

data class GitDiffFile(
    val id: UUID = UUID.randomUUID(),
    val oldPath: String,
    val newPath: String,
    val isNew: Boolean,
    val isDeleted: Boolean,
    val isRenamed: Boolean,
    val isBinary: Boolean,
    val hunks: List<GitDiffHunk>,
) {
    val displayPath: String get() = if (newPath == "/dev/null") oldPath else newPath
    val additions: Int get() = hunks.sumOf { hunk -> hunk.lines.count { it.kind == GitDiffLine.Kind.ADDITION } }
    val deletions: Int get() = hunks.sumOf { hunk -> hunk.lines.count { it.kind == GitDiffLine.Kind.DELETION } }
}

object GitDiff {
    fun parse(raw: String): List<GitDiffFile> {
        val files = mutableListOf<GitDiffFile>()
        val lines = raw.split("\n")

        var oldPath = ""
        var newPath = ""
        var isNew = false
        var isDeleted = false
        var isRenamed = false
        var isBinary = false
        var hasCurrentFile = false

        var hunks = mutableListOf<GitDiffHunk>()
        var hunkHeader: String? = null
        var hunkLines = mutableListOf<GitDiffLine>()

        fun flushHunk() {
            val header = hunkHeader
            if (header != null) {
                hunks.add(GitDiffHunk(header = header, lines = hunkLines.toList()))
            }
            hunkHeader = null
            hunkLines = mutableListOf()
        }

        fun flushFile() {
            flushHunk()
            if (hasCurrentFile) {
                files.add(
                    GitDiffFile(
                        oldPath = oldPath,
                        newPath = newPath,
                        isNew = isNew,
                        isDeleted = isDeleted,
                        isRenamed = isRenamed,
                        isBinary = isBinary,
                        hunks = hunks.toList(),
                    )
                )
            }
            hunks = mutableListOf()
            oldPath = ""
            newPath = ""
            isNew = false
            isDeleted = false
            isRenamed = false
            isBinary = false
            hasCurrentFile = false
        }

        fun stripPrefix(path: String): String =
            if (path == "/dev/null") path else path.removePrefix("a/").removePrefix("b/")

        for (line in lines) {
            when {
                line.startsWith("diff --git ") -> {
                    flushFile()
                    hasCurrentFile = true
                    val rest = line.removePrefix("diff --git ")
                    val parts = rest.split(" b/", limit = 2)
                    if (parts.size == 2) {
                        oldPath = parts[0].removePrefix("a/")
                        newPath = parts[1]
                    }
                }
                line.startsWith("new file mode") -> isNew = true
                line.startsWith("deleted file mode") -> isDeleted = true
                line.startsWith("rename from ") -> {
                    isRenamed = true
                    oldPath = line.removePrefix("rename from ")
                }
                line.startsWith("rename to ") -> {
                    isRenamed = true
                    newPath = line.removePrefix("rename to ")
                }
                line.startsWith("Binary files ") || line.startsWith("GIT binary patch") -> isBinary = true
                line.startsWith("--- ") -> oldPath = stripPrefix(line.removePrefix("--- ").trim())
                line.startsWith("+++ ") -> newPath = stripPrefix(line.removePrefix("+++ ").trim())
                line.startsWith("@@ ") -> {
                    flushHunk()
                    hunkHeader = line
                }
                hunkHeader != null && line.startsWith("+") ->
                    hunkLines.add(GitDiffLine(kind = GitDiffLine.Kind.ADDITION, text = line.removePrefix("+")))
                hunkHeader != null && line.startsWith("-") ->
                    hunkLines.add(GitDiffLine(kind = GitDiffLine.Kind.DELETION, text = line.removePrefix("-")))
                hunkHeader != null && line.startsWith("\\ No newline at end of file") -> Unit
                hunkHeader != null ->
                    hunkLines.add(GitDiffLine(kind = GitDiffLine.Kind.CONTEXT, text = line.removePrefix(" ")))
            }
        }
        flushFile()

        return files
    }
}
